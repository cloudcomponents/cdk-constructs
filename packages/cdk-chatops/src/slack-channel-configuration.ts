import { CfnSlackChannelConfiguration } from 'aws-cdk-lib/aws-chatbot';
import { Role, ServicePrincipal, PolicyStatement, Effect, ManagedPolicy } from 'aws-cdk-lib/aws-iam';
import { IFunction } from 'aws-cdk-lib/aws-lambda';
import { ITopic } from 'aws-cdk-lib/aws-sns';
import { Construct } from 'constructs';

export interface ISlackChannelConfiguration {
  readonly configurationArn: string;
}

export interface SlackChannelConfigurationProps {
  /**
   * The name of the configuration.
   */
  readonly configurationName: string;

  /**
   * The iam role that defines the permissions for AWS Chatbot.
   *
   * This is a user-defined role that AWS Chatbot will assume. This is
   * not the service-linked role. For more information, see IAM Policies
   * for AWS Chatbot.
   */
  readonly role?: Role;

  /**
   * The ID of the Slack channel.
   *
   * To get the ID, open Slack, right click on the channel name
   * in the left pane, then choose Copy Link. The channel ID is
   * the 9-character string at the end of the URL.
   * For example, ABCBBLZZZ.
   */
  readonly slackChannelId: string;

  /**
   * The ID of the Slack workspace authorized with AWS Chatbot.
   *
   * To get the workspace ID, you must perform the initial authorization
   * flow with Slack in the AWS Chatbot console. Then you can copy and
   * paste the workspace ID from the console. For more details, see steps
   * 1-4 in Setting Up AWS Chatbot with Slack in the AWS Chatbot User Guide.
   */
  readonly slackWorkspaceId: string;

  /**
   * The SNS topics that deliver notifications to AWS Chatbot.
   */
  readonly notificationTopics?: ITopic[];

  /**
   * Specifies the logging level for this configuration. This property
   * affects the log entries pushed to Amazon CloudWatch Logs.
   *
   * Logging levels include ERROR, INFO, or NONE.
   *
   * @default NONE
   */
  readonly loggingLevel?: LoggingLevel;
}

export class SlackChannelConfiguration extends Construct {
  public readonly configurationArn: string;
  public readonly role: Role;

  constructor(scope: Construct, id: string, props: SlackChannelConfigurationProps) {
    super(scope, id);

    this.role =
      props.role ||
      new Role(this, 'ServiceRole', {
        assumedBy: new ServicePrincipal('chatbot.amazonaws.com'),
      });

    const configuration = new CfnSlackChannelConfiguration(this, 'SlackChannelConfiguration', {
      configurationName: props.configurationName,
      iamRoleArn: this.role.roleArn,
      slackChannelId: props.slackChannelId,
      slackWorkspaceId: props.slackWorkspaceId,
      snsTopicArns: props.notificationTopics ? props.notificationTopics.map((topic) => topic.topicArn) : undefined,
      loggingLevel: props.loggingLevel || LoggingLevel.NONE,
    });

    this.configurationArn = configuration.ref;
  }

  /**
   * Adds a statement to the IAM role assumed by the instance.
   */
  public addToRolePolicy(statement: PolicyStatement): void {
    this.role.addToPolicy(statement);
  }

  /**
   * Allows AWS Chatbot to retreive metric graphs from Amazon Cloudwatch
   */
  public addNotificationPermissions(): void {
    this.addToRolePolicy(
      new PolicyStatement({
        effect: Effect.ALLOW,
        actions: ['cloudwatch:Describe*', 'cloudwatch:Get*', 'cloudwatch:List*'],
        resources: ['*'],
      }),
    );
  }

  public addReadOnlyCommandPermissions(): void {
    this.addToRolePolicy(
      new PolicyStatement({
        effect: Effect.DENY,
        actions: [
          'iam:*',
          's3:GetBucketPolicy',
          'ssm:*',
          'sts:*',
          'kms:*',
          'cognito-idp:GetSigningCertificate',
          'ec2:GetPasswordData',
          'ecr:GetAuthorizationToken',
          'gamelift:RequestUploadCredentials',
          'gamelift:GetInstanceAccess',
          'lightsail:DownloadDefaultKeyPair',
          'lightsail:GetInstanceAccessDetails',
          'lightsail:GetKeyPair',
          'lightsail:GetKeyPairs',
          'redshift:GetClusterCredentials',
          'storagegateway:DescribeChapCredentials',
        ],
        resources: ['*'],
      }),
    );

    this.role.addManagedPolicy(ManagedPolicy.fromAwsManagedPolicyName('ReadOnlyAccess'));
  }

  /**
   * Allows Lambda-invoke commands in supported clients
   */
  public addLambdaInvokeCommandPermissions(lambda?: IFunction): void {
    this.addToRolePolicy(
      new PolicyStatement({
        effect: Effect.ALLOW,
        actions: ['lambda:invokeAsync', 'lambda:invokeFunction'],
        resources: [lambda?.functionArn || '*'],
      }),
    );
  }

  /**
   * Allows calling AWS Support APIs in supportzed clients
   */
  public addSupportCommandPermissions(): void {
    this.role.addManagedPolicy(ManagedPolicy.fromAwsManagedPolicyName('AWSSupportAccess'));
  }
}

export enum LoggingLevel {
  ERROR = 'ERROR',
  INFO = 'INFO',
  NONE = 'NONE',
}
