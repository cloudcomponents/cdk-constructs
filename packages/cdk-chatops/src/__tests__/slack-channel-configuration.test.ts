import { Stack } from 'aws-cdk-lib';
import 'jest-cdk-snapshot';

import { SlackChannelConfiguration } from '../slack-channel-configuration';

test('default setup', (): void => {
  const stack = new Stack();

  new SlackChannelConfiguration(stack, 'SlackChannelConfiguration', {
    configurationName: 'name',
    slackChannelId: 'channelId',
    slackWorkspaceId: 'workspaceId',
  });

  expect(stack).toMatchCdkSnapshot();
});
