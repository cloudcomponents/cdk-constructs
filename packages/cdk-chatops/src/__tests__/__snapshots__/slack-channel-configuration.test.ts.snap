// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default setup 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "SlackChannelConfiguration9AFE1994": Object {
      "Properties": Object {
        "ConfigurationName": "name",
        "IamRoleArn": Object {
          "Fn::GetAtt": Array [
            "SlackChannelConfigurationServiceRole517C6CD4",
            "Arn",
          ],
        },
        "LoggingLevel": "NONE",
        "SlackChannelId": "channelId",
        "SlackWorkspaceId": "workspaceId",
      },
      "Type": "AWS::Chatbot::SlackChannelConfiguration",
    },
    "SlackChannelConfigurationServiceRole517C6CD4": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "chatbot.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
